---
title: "ContrApption"
output: rmarkdown::html_notebook
---

ContrApption allows users to easily create an interactive JavaScript widget that accepts user input via a searchable drop-down menu and produces boxplot visualizations of targets in RNA-Seq and similarly structured datasets. This permits easy, visual inspection of results from a single function call in R and, because ContrApption is written in JavaScript and runs in a browser, sharing of those results without hosting a Shiny server. For the most basic usage, ContrApption simply requires:

  * A matrix or data.frame of numeric data to visualize where the columns are the samples in the experiment ("numeric data").
  * A matrix or data.frame of meta describing the experimental features of the experiment and which samples
  belong to which experimental conditions ("annotation data").
   
Note that the row names of the annotation data must mach the column names of the numeric data. This reflects the requirements of the RNA-Seq
experiments for which ContrApption was designed. An intuitive visualization of this data structure can be found on page 5 of the [Beginnerâ€™s guide to using the DESeq2 package](https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf). There is no reason a user from other domains with data in the same structure cannot use the tool however, as the inputs required are simply strings and numbers.

For more complex usage, the user may also provide differential expression data and create a shared data object so the dataset can be visualized in the crosstalk-compatible [DT](https://rstudio.github.io/DT/) widget and ContrApption simultaneously.

In this vignette we will walk through simple and complex usage of ContrApption with the [pasilla dataset](https://bioconductor.org/packages/release/data/experiment/html/pasilla.html) used in the [DESeq2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

## Vignette Dependencies

We load the requires libraries:

```{r setup}
devtools::install()
library(ContrApption)
library(tidyverse)

# libraries for optional integration
library(DT)
library(crosstalk)

suppressMessages(
  {
    library(DESeq2)
    library(edgeR)
    library(limma)
    library(pasilla)
  }
)
```

## Dataset

This vignette makes use of [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and the [pasilla dataset](https://bioconductor.org/packages/release/data/experiment/html/pasilla.html) to demonstrate the intended use of ContrApption. The following is a modified version of the ["Count matrix input"](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#quick-start) section of the [DESeq2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 

### Annotation Data and Groups

The following snippet retrieves the column metadata for the Pasilla experiment.

```{R}

# loads sample annotation from the pasilla package
pasAnno <- system.file("extdata", "pasilla_sample_annotation.csv",
                       package = "pasilla", mustWork = TRUE)

# read in the sample data
coldata <- read.csv(pasAnno, row.names = 1)

# select relevant columns
coldata <- data.frame(coldata[ , c("condition","type")])

# remove un-needed characters
rownames(coldata) <- gsub("fb", "", rownames(coldata))

# visualize our metadata
coldata
```

This matrix will be our `annotation` input. Here the rownames are the sample names of the experiment.

### Numeric Data

The Pasilla dataset contains per-exon and per-gene read counts of RNA-seq samples which will be our `data` input. The exons/genes are the rows, and the samples in the experiment are the columns. The following code extracts the counts from the package.

```{R}

# loads counts file from the pasilla package
pasCts <- system.file("extdata", "pasilla_gene_counts.tsv",
                      package = "pasilla", mustWork = TRUE)

# reads counts of genes
cts <- read.csv(pasCts, sep = "\t", row.names = "gene_id")

# make sure the order is correct
cts <- cts[, rownames(coldata)]

# examine the counts
head(cts)
```

We now have all we need to run DESeq2 and thus to run ContrApption.

## DESeq2 Normalized Counts Example

We can create a DESeq2 dataset, run DESeq2 on it, and then extract the normalized counts for visualization.

```{R}

# create a DESeq2 dataset from the metadata and counts
dds <- DESeq(DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ condition))

# extract the normalized counts
normCounts <- counts(dds, normalized = TRUE)

head(normCounts)
```


Given that embedding multiples widgets with 14599 rows of data may lead to performance issues, the user may want to focus one genes of interest as determined by the differential expression experiment:

```{R}

# extract significant results
res <- results(dds, tidy = TRUE) %>% 
  data.frame %>% 
  filter(padj < 0.05)

# get the counts for signficant transcripts
normCountsSig <- normCounts %>%
  data.frame %>% 
  filter(rownames(.) %in% res$row)

normCountsSig %>% nrow
```
```{R}

# create a column for out targets
normCountsSig$geneToob <- rownames(normCountsSig)

# create a shared data object
normCountsSigShared <- SharedData$new(normCountsSig)
```

For organizational purposes, We pass the two widgets to the `bscols` function from the "crosstalk" package, and then specify the `scaleWidth` argument to ContrApption so it will resize appropriately:

```{R}

# # creates a side-by-side pair of widgets
pairedWidgetsCounts <- bscols(

  # the datatable widget
  datatable(
    normCountsSigShared,
    extensions = "Scroller",
    style = "bootstrap",
    class = "compact",
    width = "100%",
    selection = 'single', # ContrApption works on one target at a time
    options = list(
      deferRender = TRUE,
      scrollY = 300,
      scroller = TRUE
    )
  ),

  # the ContrApption widget
  ContrApption(
    data = normCountsSigShared,
    annotation = coldata,
    scaleWidth = 0.5, # makes room for other widgets,
    targetCol = "geneToob"
  )

)

pairedWidgetsCounts
```
