---
title: "ContrApption"
output: rmarkdown::html_notebook
---

ContrApption allows users to easily create an interactive JavaScript widget that accepts user input via a searchable drop-down menu and produces boxplot visualizations of targets in RNA-Seq and similarly structured datasets. This permits easy, visual inspection of results from a single function call in R and, because ContrApption is written in JavaScript and runs in a browser, sharing of those results without hosting a Shiny server. For the most basic usage, ContrApption simply requires:

  * A data.frame of numeric data to visualize where the columns are the samples in the experiment ("numeric data").
  * A data.frame of metadata describing the experimental features and denoting which samples
  belong to which experimental conditions ("annotation data").
   
Note that the row names of the annotation data must mach the column names of the numeric data. This reflects the requirements of the RNA-Seq
experiments for which ContrApption was designed. An intuitive visualization of this data structure can be found on page 5 of the [Beginnerâ€™s guide to using the DESeq2 package](https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf). There is no reason a user from other domains with data in the same structure cannot use the tool however, as the inputs required are simply strings and numbers.

For more complex usage, the user may create a shared data object from the counts dataframe so it can be visualized in ContrApption and in the [DT](https://rstudio.github.io/DT/) datatables widget simultaneously. Selections in one widget will update the other.  Furthermore, the user may provide differential expression data and counts data so that selections in a table of differential expression results can be plotted live in ContrApption as they are explored. 

In this vignette we will walk through simple and complex usage of ContrApption with the [pasilla dataset](https://bioconductor.org/packages/release/data/experiment/html/pasilla.html) used in the [DESeq2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html).

## Vignette Dependencies

We load the requires libraries:

```{r setup}
devtools::install()
library(ContrApption)
library(tidyverse)

# libraries for optional integration
library(DT)
library(crosstalk)

suppressMessages(
  {
    library(DESeq2)
    library(edgeR)
    library(limma)
    library(pasilla)
  }
)
```

## Dataset

This vignette makes use of [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and the [pasilla dataset](https://bioconductor.org/packages/release/data/experiment/html/pasilla.html) to demonstrate the intended use of ContrApption. The following is a modified version of the ["Count matrix input"](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#quick-start) section of the [DESeq2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 

### Annotation Data and Groups

The following snippet retrieves the column metadata for the Pasilla experiment.

```{R}

# loads sample annotation from the pasilla package
pasAnno <- system.file("extdata", "pasilla_sample_annotation.csv",
                       package = "pasilla", mustWork = TRUE)

# read in the sample data
coldata <- read.csv(pasAnno, row.names = 1)

# select relevant columns
coldata <- data.frame(coldata[ , c("condition","type")])

# remove un-needed characters
rownames(coldata) <- gsub("fb", "", rownames(coldata))

# visualize our metadata
coldata
```

This matrix will be our `annotation` input. Here the rownames are the sample names of the experiment.

### Numeric Data

The Pasilla dataset contains per-exon and per-gene read counts of RNA-seq samples which will be our `data` input. The exons/genes are the rows, and the samples in the experiment are the columns. The following code extracts the counts from the package.

```{R}

# loads counts file from the pasilla package
pasCts <- system.file("extdata", "pasilla_gene_counts.tsv",
                      package = "pasilla", mustWork = TRUE)

# reads counts of genes
cts <- read.csv(pasCts, sep = "\t", row.names = "gene_id")

# make sure the order is correct
cts <- cts[, rownames(coldata)]

# examine the counts
head(cts)
```

We now have all we need to run DESeq2 and thus to run ContrApption.

## DESeq2 Normalized Counts Example

We can create a DESeq2 dataset, run DESeq2 on it, and then extract the normalized counts for visualization.

```{R}

# create a DESeq2 dataset from the metadata and counts
dds <- DESeq(DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ condition))

# extract the normalized counts
normCounts <- counts(dds, normalized = TRUE)

head(normCounts)
```


Given that embedding multiples widgets with 14599 rows of data may lead to performance issues, the user may want to focus one genes of interest as determined by the differential expression experiment:

```{R}

# extract significant results
res <- results(dds, tidy = TRUE) %>% 
  data.frame %>% 
  filter(padj < 0.05)

# get the counts for signficant transcripts
normCountsSig <- normCounts %>%
  data.frame %>% 
  filter(rownames(.) %in% res$row) %>% 
  format(digits = 2)

normCountsSig %>% nrow
```

When used without crosstalk, ContrApption will look for rownnames to determine the list of targets of the targets. If there is an existing column the user would like to use, they can change this behavior by specifying a `targetCol` argument. Similarly, the the sample IDs are assumed to be the rownames of the annotation file, but the user may specify a column using the `sampleCol` argument.

## Basic Usage

We'll demonstrate the most straightforward use cases of ContrApption - visualizing count data and using pre-defined functions to leverage crosstalk compatibility.

### Basic counts usage

The most basic usage of ContrApption is to pass counts and annotations to it. We pass the dataset and the annotation file to ContrApption, resulting in a widget with a searchable drop-down menu of experimental features and boxplots of the input data grouped as directed. 

```{R}

# the basic usage of ContrApption
ContrApption(data = normCountsSig, annotation = coldata)
```

Users may customize the title and y-axis labels of the plot by setting the `plotName` and `yAxisName` arguments.

```{R}

ContrApption(
  data = normCountsSig,
  annotation = coldata,
  plotName = "DESeq2 normalized counts by sample types",
  yAxisName = "norm. counts"
)
```

### Interaction-style example

ContrApption operates on experimental factors with an arbitrary numbers of levels. We can leverage this to explore the interaction of multiple factors of interest by merging the columns of the annotation file.

```{R}

# derive a new variable - every combination of condition and type
coldata$interaction <- paste0(coldata$condition, ":", coldata$type)

head(coldata)
```

We can then use the function normally, and we'll se an extra option in the dropdown menu that allows us to explore our new variable. 

```{R}

# pass our new variable
ContrApption(data = normCountsSig, annotation = coldata)
```

    Note that users interested in this sort of experiment would generally re-run DESeq2 with a different experimental design: 
    >DESeq(DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ condition * type))

## Crosstalk Integration

### Counts

ContrApption is [crosstalk](https://rstudio.github.io/crosstalk/index.html) compatible for usage with [DT](https://rstudio.github.io/DT/). Users may leverage this by creating a shared data object and passing it to the `datatables` and `ContrApption` functions to produce interactive widgets that react to inputs in one another. Selecting a row in the datatable will result in a call to visualize the selected row in ContrApption, and selecting a target in ContrApption will filter the table to that row in the datatable. ContrApption will not changed an object once it becomes a shared data object, and they do no have rownames, so when using crosstalk, create the "gene" column before creating the shared data object.

```{R}

# create a column for out targets
normCountsSig$gene <- rownames(normCountsSig)

# create a shared data object
normCountsSigShared <- SharedData$new(normCountsSig)
```

We're  now ready to use the cross-enabled function of ContrApption.

#### Paried-widget counts function

The `pairedCountsWidget` will create a side-by-side pair of widgets - one showing the counts and one showing the ContrApption visualization. It simply takes the same `data` and `annotation` arguments as in the basic usage above, but data is now a shared object.

```{R}

pairedCountsWidget(data = normCountsSigShared, annotation = coldata)
```
```{R}

pairedWidget(counts = normCountsSigShared, annotation = coldata)
```
### Differential Expression Data

Users may also be interested in viewing the results of tests for differential expression with the visualizations provided. This can be achieved by creating a shared data object from the results table and passing `"diff-expression"` to the `mode` argument of ContrApption. This disables the dropdown for the target column in ContrApption as it is no longer needed when the user has a searchable table. The usage is similar to the above, the shared object is created from the expression data (as this is the data that both widgets need, only ContrApption will need the counts), and counts are passed as supplemental information using the `counts` argument. Also note that the `gene`, or other target name column must be created before passing the data to `SharedData$new`.

```{R}

normCountsSig <- normCounts %>%
  data.frame %>% 
  filter(rownames(.) %in% res$row)

normCountsSig$gene <- rownames(normCountsSig)

# get the significant targets from the DESeq2 experiment and format them
res <- results(dds) %>% 
  data.frame %>% 
  dplyr::filter(padj < 0.05) %>% 
  format(digits = 2)

resShared <- SharedData$new(res)
```

#### Paried-widget differential expression function

The DE version of the paired widget takes the following arguments: `deResults` for the DE table, `counts` for the counts, and `annotation` for the annotation/column data. 

```{R}

pairedDiffExWidget(deResults = resShared, counts = normCountsSig, annotation = coldata)
```

```{R}

pairedWidget(counts = normCountsSig, deResults = resShared, annotation = coldata)
```
## Advanced Usage

This section of the vignette demonstrates the methods used in the ContrApption internals to leverage `datatables` for users interested in creating customized widgets.

### Customizing the paired widget functions

Users interested in customizing the paired widgets are encouraged to visit the documentation for [DT](https://rstudio.github.io/DT/) and [DataTables](https://www.datatables.net/), the JavaScript library it wraps, to learn to of the numerous options it provides and how they can be passed to `crosstalk` for use with `ContrApption`.

#### Counts

This block demonstrates how the `pairedCountsWidget` function works. A customized `datatables` widget is made, and passed to `crosstalk`'s `bscols` function along with a `ContrApption` widget to create a pair.

```{R}

# first up, a fairly extensively customized datatables object
dtCountsWidget <-  datatable(
  normCountsSigShared,     # Our shared data
  extensions = "Scroller", # Various tweaks to the table described in the DT docs 
  style = "bootstrap",     
  class = "compact",
  width = "100%",
  selection = 'single', # ContrApption works on one transcript at at time, we we only need one
  # This is a list of options that are passed to the DataTables, described in the DataTables docs
  options = list(
    deferRender = TRUE,
    scroller = TRUE,
    scrollY = 300,
    sScrollX = "100%",
    pagingType = "simple",
    # This option allows the user to pass raw, arbitrary JavaScript to be executed when the table
    # is rendered. proceed with caution.
    initComplete = JS(
      "function(settings, json) {",
        "$(this.api().table().header()).css({'font-size': '75%'});",
      "}"
    )
  )
  # This is a convenience function to add style tweaks to the table. In this case, 
  # the fontsize of columns is changed on columns listed by number (all of them here)
) %>% formatStyle(columns = seq(1, ncol(normCountsSigShared$origData())), fontSize = '75%')

# ContrApption, used in a fairly basic way, scaled down 50%
cntrCountsWidget <- ContrApption(
  data = normCountsSigShared,
  annotation = coldata,
  scaleWidth = 0.5 # makes room for other widgets,
)

# creates a side-by-side pair of widgets using bscols
pairedWidgetsCounts <- bscols(dtCountsWidget, cntrCountsWidget)

pairedWidgetsCounts
```

#### Differential Expression

This example show the same as above, with minor modification to incorporate DE results. Most notable, the shared object is created from the expression data, and counts are passed as supplemental information using the `counts` argument.

```{R}

# create a datatables widget
dtDEwidget <- datatable(
  resShared,
  extensions = "Scroller",
  style = "bootstrap",
  class = "compact",
  width = "100%",
  selection = 'single',
  options = list(
    deferRender = TRUE,
    scroller = TRUE,
    scrollY = 300,
    sScrollX = "100%",
    pagingType = "simple",
    initComplete = JS(
      "function(settings, json) {",
        "$(this.api().table().header()).css({'font-size': '75%'});",
      "}"
    )
  )
) %>% formatStyle(columns = seq(1, ncol(resShared$origData())), fontSize = '75%')

# create a ContrApption widget
cDEwidget <- ContrApption(
  data = resShared,
  countsData = normCountsSig,
  annotation = coldata,
  mode = "diff-expression",
  scaleWidth = 0.5
)

pairedWidgetsDE <- bscols(dtDEwidget, cDEwidget)

pairedWidgetsDE
```

## Misc

This section demonstrates additional examples in slightly different contexts and the use of accesory arguments.

### Example with limma-voom results

This section shows a basic [limma-voom](https://bioconductor.org/packages/release/bioc/html/limma.html) DE analysis the ends in exploration of the results using `ContrApption`.

#### Basic

We conduct a basic `limma-voom` analysis:

```{R}

# the experimental design matrix
design <- model.matrix(~1 + condition, coldata)

# edgeR DE list
dge <- DGEList(counts = as.matrix(cts))

# created index of non-expressed for filtration
toKeep <- filterByExpr(dge, design)

# filter out non-expressed
dge <- dge[toKeep, keep.lib.sizes = FALSE]

# calc normnalization factors
dge <- calcNormFactors(dge)

# add voom adjustment
voomObj <- voom(dge, design, plot = FALSE)

# fit the model
fit <- lmFit(voomObj, design)

# calculate tests statistics
fit <- eBayes(fit)

# return the results
resLV <- topTable(fit, n = Inf, coef = NULL) %>% filter(adj.P.Val < 0.05) %>%
  format(digits = 2)

# "normalized counts" https://support.bioconductor.org/p/103747/#103769
normCountsLV <- cpm(dge, log = TRUE, prior.count = 3) %>% 
  format(digits = 2)

normCountsSigLV <- normCountsLV %>%
  data.frame %>%
  dplyr::filter(rownames(.) %in% rownames(resLV))
```

We pass the results to `ContrApption`:

```{R}

ContrApption(data = normCountsSigLV, annotation = coldata)
```

#### Interactive Counts

An example with the interactive counts, customized explicitly:

```{R}

normCountsSigLV$gene <- rownames(normCountsSigLV)

normCountsSigLVShared <- SharedData$new(normCountsSigLV)

dtDEwidgetLV <- datatable(
  normCountsSigLVShared,
  extensions = "Scroller",
  style = "bootstrap",
  class = "compact",
  width = "100%",
  selection = 'single',
  options = list(
    deferRender = TRUE,
    scroller = TRUE,
    scrollY = 300,
    sScrollX = "100%",
    pagingType = "simple",
    initComplete = JS(
      "function(settings, json) {",
        "$(this.api().table().header()).css({'font-size': '75%'});",
      "}"
    )
  )
) %>% formatStyle(columns = seq(1, ncol(normCountsSigLVShared$origData())), fontSize = '75%')

cntrDEwidgetLV <- ContrApption(
  data = normCountsSigLVShared, # new shared results
  annotation = coldata,
  scaleWidth = 0.5
)

pairedWidgetsDELV <- bscols(dtDEwidgetLV, cntrDEwidgetLV)

pairedWidgetsDELV
```

#### Differential Expression

An example with the interactive DE, customized explicitly:

```{R}

resLVShared <-  SharedData$new(resLV)

dDiffExWidgetLV <- datatable(
  resLVShared,
  extensions = "Scroller",
  style = "bootstrap",
  class = "compact",
  width = "100%",
  selection = 'single',
  options = list(
    deferRender = TRUE,
    scroller = TRUE,
    scrollY = 300,
    sScrollX = "100%",
    pagingType = "simple",
    initComplete = JS(
      "function(settings, json) {",
        "$(this.api().table().header()).css({'font-size': '75%'});",
      "}"
    )
  )
) %>% formatStyle(columns = seq(1, ncol(resLVShared$origData())), fontSize = '75%')

cntrDiffExWidgetLV <- ContrApption(
  data = resLVShared,
  countsData = normCountsSigLV,
  annotation = coldata,
  mode = "diff-expression",
  scaleWidth = 0.5
)

pairedWidgetsDELV <- bscols(dDiffExWidgetLV, cntrDiffExWidgetLV)

pairedWidgetsDELV
```

### Specifying `targetCol`

Using a user-specified column instead of the implicit rownames.

```{R}

# remove gene column
normCountsSig$gene <- NULL

# make our own
normCountsSig$ArbitraryString <- rownames(normCountsSig)

ContrApption(data = normCountsSig, annotation = coldata, targetCol = "ArbitraryString")
```

### Specifying `sampleCol`

Using a user-specified column instead of the implicit rownames.

```{R}

# remove gene column
normCountsSig$ArbitraryString <- NULL

# make a column in the annotation matrix
coldata$MyCustomID <- rownames(coldata)

# remove rownames
rownames(coldata) <- NULL

ContrApption(data = normCountsSig, annotation = coldata, sampleCol = "MyCustomID")
```

```{R}

sessionInfo()
``` 
