---
title: "ContrApption"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ContrApption}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```



ContrApption allows users to easily create a boxplot that responds to user input via a dropdown menu without using JavaScript and with minimal R usage. It requires 3 inputs:

  * A matrix or data.frame of numeric data to visualize where the columns are the samples in the experiment ("numeric data").
  * A matrix or data.frame of meta describing the experimental features of the experiment and which samples
  belong to which experimental conditions ("annotation data").
  * The name of an experimental factor (as a string) by which to group the input data
    
Note that the row names of the annotation data must mach the column names of the numeric data. This is reflect the requirements of the RNA-Seq
experiments for which ContrApption was designed. An intuitive visualization of this data structure can be found on page 5 of the [Beginnerâ€™s guide to using the DESeq2 package](https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf). There is no reason a users from other domains with data in the same structure cannot use the tool however, as the inputs required are simply strings and numbers.

## Dependencies

We load the requires libraries:

```{r setup}

library(ContrApption)
library(dplyr)

suppressMessages(library(DESeq2))
suppressMessages(library(pasilla))
```

## Dataset

This vignette makes use of [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) and the [pasilla dataset](https://bioconductor.org/packages/release/data/experiment/html/pasilla.html) to demonstrate the intended use of ContrApption. The following is a modified version of the ["Count matrix input"](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#quick-start) section of the [DESeq2 vignette](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html) 

### Annotation Data and Groups

The following snippet retrieves the column metadata for the Pasilla experiment.

```{R}

# loads sample annotation from the pasilla package
pasAnno <- system.file("extdata", "pasilla_sample_annotation.csv", package = "pasilla", mustWork = TRUE)

# read in the sample data
coldata <- read.csv(pasAnno, row.names = 1)

# select relevant columns
coldata <- data.frame(coldata[ , c("condition","type")])

# remove un-needed characters
rownames(coldata) <- gsub("fb", "", rownames(coldata))

coldata
```

This matrix will be our `annotation` input. Here the rownames are the sample names of the experiment and either 'condition' or 'type' may be used as a group for `groupCol`.

### Numeric Data

The Pasilla dataset contains per-exon and per-gene read counts of RNA-seq samples which will be our `data` input. The exons/genes are the rows, and the samples in the experiment are the columns. The following code extracts the counts from the pckage.

```{R}

# loads counts file from the pasilla package 
pasCts <- system.file("extdata", "pasilla_gene_counts.tsv", package = "pasilla", mustWork = TRUE)

# reads counts of genes
cts <- read.csv(pasCts, sep = "\t", row.names = "gene_id")

# make sure the order is correct
cts <- cts[, rownames(coldata)]

head(cts)
```

### DESeq2

We can create a DESeq2 dataset and then extract the normalized counts for visualization.

```{R}

# create a DESeq2 dataset from the metadata and counts
dds <- DESeq(DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ condition))

normCounts <- counts(dds, normalized = TRUE)
```

Given that embedding multiples widgets with 14599 rows of data may lead to performance issues, the user may want to focus one genes of interest as determined by the differential expression experiment:

```{R}

res <- results(dds, tidy = TRUE) %>% 
  data.frame %>% 
  filter(padj < 0.05)

normCountsSig <- normCounts %>%
  data.frame %>% 
  filter(rownames(.) %in% res$row)
  
normCountsSig %>% nrow
```

## Examples

### Two-factor Examples

We pass the dataset, the annotation file, and the column contaning the experimental factor of interest to ContrApption, resulting in a widget with a dropdown menu of features and boxplots of the input data grouped as directed. 

```{R}

ContrApption(data = normCountsSig, annotation = coldata, groupCol = "condition")
```


Users may create multiple widgets per book. In this experiment there is more than one experimental factor, so we can create another widget based on a different group. We may also use the optional y-axis label anf plot title arguments argument.

```{R}

ContrApption(
  data = normCountsSig,
  annotation = coldata,
  groupCol = "type",
  plotName = "DESeq2 normalized counts by sample types",
  yAxisName = "counts"
)
```

### Interaction-style example

ContrApption operates on experimental factors with an arbitrary numbers of levels. We can use this to subset by multiple factors of interest but merging the columns of the annotation file.

```{R}

# derive a new variable - every combitation of condition and type
coldata$interaction <- paste0(coldata$condition, ":", coldata$type)

head(coldata)
```

We can then use the function normally:

```{R}

# pass our new variable
ContrApption(data = normCountsSig, annotation = coldata, groupCol = "interaction")
```

    Note that users interested in this sort of experiment would generally re-run DESeq2 with a different experimental design: 
    >DESeq(DESeqDataSetFromMatrix(countData = cts, colData = coldata, design = ~ condition * type))

```{R}

sessionInfo()
```

